qa/L0_backend_python/response_sender/response_sender_test.py

```diff
@@ -538,10 +542,62 @@ def test_decoupled_one_response_on_and_post_return(self):
         # TODO: Test for async decoupled after fixing 'AsyncEventFutureDoneCallback'
         #       using `py_future.result()` with error hangs on exit.
 
+    # Non-decoupled model send response final flag before request return.
+    def test_non_decoupled_zero_response_pre_return(self):
+        # Note: The final flag will raise an exception which stops the model. Since the
+        #       exception happens before the model returns, it will be caught by the
+        #       stub process which pass it to the backend and sent an error response
+        #       with final flag.
+        number_of_response_before_return = 0
+        send_complete_final_flag_before_return = True
+        return_a_response = False
+        number_of_response_after_return = 0
+        send_complete_final_flag_after_return = False
+        expected_message = (
+            "Non-decoupled model cannot send complete final before sending a response"
+        )
+
+        model_name = "response_sender"
+        responses = self._infer(
+            model_name,
+            number_of_response_before_return,
+            send_complete_final_flag_before_return,
+            return_a_response,
+            number_of_response_after_return,
+            send_complete_final_flag_after_return,
+        )
+        self._assert_responses_exception(responses, expected_message)
+        # Do NOT group into a for-loop as it hides which model failed.
+        model_name = "response_sender_async"
+        responses = self._infer(
+            model_name,
+            number_of_response_before_return,
+            send_complete_final_flag_before_return,
+            return_a_response,
+            number_of_response_after_return,
+            send_complete_final_flag_after_return,
+        )
+        self._assert_responses_exception(responses, expected_message)
+
+    # Non-decoupled model send response final flag after request return.
+    @unittest.skip("Model unload will hang, see the TODO comment.")
+    def test_non_decoupled_zero_response_post_return(self):
```

Comment:
By "hang" do you mean the timeout counter that starts when the request is getting unloaded? If that's the case, I think it is fine since closing the response sender failed so they need to fix their model.

Reply:
> By "hang" do you mean the timeout counter that starts when the ~~request~~ **(model?)** is getting unloaded?

Yes, the hang happens when the server is gracefully shutting down. The model should not raise an exception an any point (or at least catch the exception itself).